### ISSUE TITLE
Can Mitiq support circuit type <class 'pennylane.tape.tape.QuantumTape'>?

### ISSUE THREAD

**Issue by eegzheng** [unitaryfund/mitiq#1694]
## Proposed Solution

I wonder whether Mitiq can support circuit type <class 'pennylane.tape.tape.QuantumTape'> such as that used in this Pennylane example https://pennylane.ai/qml/demos/tutorial_quantum_transfer_learning.html?

## Additional References

I tried the error mitigation method https://pennylane.ai/blog/2021/11/error-mitigation-with-mitiq-and-pennylane/ on the circuit in the above link but got the error that "Circuit could not be converted to an internal Mitiq circuit. Provided circuit has type <class 'pennylane.tape.tape.QuantumTape'>". This seems a quite standard circuit, so it would be great if Mitiq can support error mitigation of it.

**Comment by andreamari** [comment]
Hi @eegzheng, thanks for posting this!

The error message is a bit misleading. Mitiq is not able to convert some gates into its internal representation.  In particular Mitiq doesn't support gates with symbolic parameters.
This is probably the reason why Mitiq is not working in your example. 

Variational circuits are ok but parameters must be real numbers not symbolic variables.  In principle, this is not a strong limitation since Mitiq is supposed to be applied at the end of the execution stack, where parameters are usually replaced by real numbers before going into the backend. 

However, in practice, I understand that this can be a blocker if one wants to easily apply Mitiq to some code heavily based on symbolic parameters.

**Comment by eegzheng** [comment]
Dear Andreal, many thanks for your kind reply. I wanted to see whether Mitiq can be used for the transfer learning example you provided https://pennylane.ai/qml/demos/tutorial_quantum_transfer_learning.html, or if this is not possible for training, is it possible for inference at least? Are gate parameters in your transfer example symbolic variables? Thanks.

**Comment by andreamari** [comment]
Yes, gate parameters are symbolic in the transfer learning example. Honestly I don't know if Mitiq can be applied in this kind of scenario. But I have some suggestions in the case you want to try:

- Focus first on the inference phase as you mentioned, no training. In principle this is bad, since training and inference should happen at the same noise level, but it is a first step to understand if there are blockers.
- Try first to decorate the qnode with the PennyLane native decorator for error mitigation, following [this part](https://pennylane.ai/qml/demos/tutorial_error_mitigation.html#error-mitigation-in-pennylane) of the PennyLane guide. If this doesn't work, see next step.
- You can also try to dynamically define a quantum tape depending on the value of some real numerical parameters. The parameters should not be arguments of the tape/node. Just numbers. I am not sure how this can be implemented in the transfer learning example, but a different example in which this approach works is shown in [this part](https://pennylane.ai/qml/demos/tutorial_error_mitigation.html#mitigating-noisy-circuits-in-quantum-chemistry) of the PennyLane guide. If this doesn't work, I don't know how to help :-)

**Comment by Misty-W** [comment]
Follow up: reproduce issue and possibly improve error message.

**Comment by andreamari** [comment]
Hi @eegzheng,

I did some tests that may help better understand your issue.  From the test snippet attached below it seems that:
1. The simple circuit in the transfer learning example seems compatible with Mitiq, at least in the case in which you don't need to compute the gradient (no training). I wasn't able to reproduce the  `Circuit could not be converted to an internal Mitiq circuit. Provided circuit has type <class 'pennylane.tape.tape.QuantumTape'>"` with my code snippet. 
2. The same simple circuit seems compatible with native PennyLane error mitigation functions. In this case, the mitigated qnode seems trainable too (see final `requires_grad=True` printed output of the code snippet).

I didn't test the same with pytorch since I don't have it installed in my machine. But a tentative answer to your question is the following:

- Yes, Mitiq can be applied to the PL qnode as long as you don't need to evaluate gradients.
- If you instead need to evaluate gradients, it is probably necessary to use built-in functions (`fold_global`, `richardson_extrapolate`) from `pennylane.transforms`.

## Code snippet with ZNE applied to the transfer learning quantum circuit
<details>
  <summary>Click to expand!</summary>
  
```python
import pennylane as qml
from pennylane import device as dev
import numpy as np


# Mitigation packages
from pennylane.transforms import mitigate_with_zne

# ZNE packages from Mitiq
from mitiq import zne
from mitiq.zne.scaling import fold_global
from mitiq.zne.inference import RichardsonFactory

# ZNE packages from from PL
from pennylane.transforms import fold_global as pennylane_fold_global
from pennylane.transforms import richardson_extrapolate


n_qubits = 4                # Number of qubits
q_depth = 6                 # Depth of the quantum circuit (number of variational layers)

dev = qml.device("default.qubit", wires=n_qubits)


def H_layer(nqubits):
    """Layer of single-qubit Hadamard gates.
    """
    for idx in range(nqubits):
        qml.Hadamard(wires=idx)


def RY_layer(w):
    """Layer of parametrized qubit rotations around the y axis.
    """
    for idx, element in enumerate(w):
        qml.RY(element, wires=idx)


def entangling_layer(nqubits):
    """Layer of CNOTs followed by another shifted layer of CNOT.
    """
    # In other words it should apply something like :
    # CNOT  CNOT  CNOT  CNOT...  CNOT
    #   CNOT  CNOT  CNOT...  CNOT
    for i in range(0, nqubits - 1, 2):  # Loop over even indices: i=0,2,...N-2
        qml.CNOT(wires=[i, i + 1])
    for i in range(1, nqubits - 1, 2):  # Loop over odd indices:  i=1,3,...N-3
        qml.CNOT(wires=[i, i + 1])
        
@qml.qnode(dev)
def quantum_net(q_input_features, q_weights_flat):
    """
    The variational quantum circuit.
    """
    # Reshape weights
    q_weights = q_weights_flat.reshape(q_depth, n_qubits)

    # Start from state |+> , unbiased w.r.t. |0> and |1>
    H_layer(n_qubits)

    # Embed features in the quantum node
    RY_layer(q_input_features)

    # Sequence of trainable variational layers
    for k in range(q_depth):
        entangling_layer(n_qubits)
        RY_layer(q_weights[k])

    # Expectation values in the Z basis
    exp_vals = [qml.expval(qml.PauliZ(position)) for position in range(n_qubits)]
    return tuple(exp_vals)

#### Mitigate with Mitiq ###########
folding_function = fold_global
extrapolate = RichardsonFactory.extrapolate
scale_factors = [1, 3]
mitigated_qnode = mitigate_with_zne(scale_factors, folding_function, extrapolate)(quantum_net)


##### Mitigate with PL  ############
folding_function_pl = pennylane_fold_global
extrapolate_pl = richardson_extrapolate
scale_factors = [1, 3]
mitigated_qnode_pl = mitigate_with_zne(scale_factors, folding_function_pl, extrapolate_pl)(quantum_net)

in_data = np.array([1.0 for _ in range(4)])
w_data = np.array([1.0 for _ in range(24)])

print("Unmitigated qnode:")
print(quantum_net)
print(repr(quantum_net(q_input_features=in_data, q_weights_flat=w_data)))
print()

print("Mitigated with Mitiq:")
print(mitigated_qnode)
print(repr(mitigated_qnode(q_input_features=in_data, q_weights_flat=w_data)))
print()


print("Mitigated with PL:")
print(mitigated_qnode_pl)
print(repr(mitigated_qnode_pl(q_input_features=in_data, q_weights_flat=w_data)))
```
PRINTED OUTPUT:

```
Unmitigated qnode:
<QNode: wires=4, device='default.qubit.autograd', interface='autograd', diff_method='best'>
tensor([-0.59838279,  0.57432842, -0.80800684,  0.37394465], requires_grad=True)

Mitigated with Mitiq:
<function quantum_net at 0x7f4a83f7b400>
array([-0.59838279,  0.57432842, -0.80800684,  0.37394465])

Mitigated with PL:
<function quantum_net at 0x7f4a83f7b1c0>
tensor([-0.59838279,  0.57432842, -0.80800684,  0.37394465], requires_grad=True)
```
</details>

**Comment by Misty-W** [comment]
Closing because, as noted above, Mitiq can be applied to the PL qnode as long as you don't need to evaluate gradients.
If you instead need to evaluate gradients, it is probably necessary to use built-in functions (fold_global, richardson_extrapolate) from pennylane.transforms.

**Comment by eegzheng** [comment]
Hi Andrea, thanks for investigating this further. It is true that the simple quantum circuit used in the transfer learning example is compatible with Mitiq and PL error mitigation; however, when you add the classical layers as that in the transfer learning, then the error I reported at the beginning will appear.

I managed to make error mitigation work for transfer learning only for inference, by separating the quantum circuit and the classical layers, e.g., in Pytorch, taking the output of middle layer/feed quantum output to classical layer. For training hybrid neural network, error mitigation still does not work.

**Comment by andreamari** [comment]
Thanks @eegzheng, for the update! What you are reporting seems consistent with the mentioned impossibility of computing gradients when using Mitiq+PL. Maybe training could work when using the native error mitigation tools in `pennylane.transforms` instead of Mitiq, but I am not familiar with them.
